rust/xaynet-core/src/mask/scalar.rs

```diff
@@ -0,0 +1,250 @@
+            target: PrimitiveType::I64,
+        })
+    }
+
+    fn to_primitive(&self) -> Result<i64, ScalarCastError> {
+        let i = self.0.clone();
+        i.to_integer().to_i64().ok_or(ScalarCastError {
+            weight: i,
+            target: PrimitiveType::I64,
+        })
+    }
+}
+
+impl FromPrimitive<i64> for Scalar {
+    fn from_primitive(prim: i64) -> Result<Self, PrimitiveCastError<i64>> {
+        let i = BigUint::try_from(prim).map_err(|_| PrimitiveCastError(prim))?;
+        Ok(Self(Ratio::from_integer(i)))
+    }
+
+    fn from_primitive_bounded(prim: i64) -> Self {
+        Self::from_primitive(prim).unwrap_or_else(|_| Self(Ratio::zero()))
+    }
+}
+
+impl IntoPrimitive<f32> for Scalar {
+    fn into_primitive(self) -> Result<f32, ScalarCastError> {
+        let r = self.to_ratio();
+        ratio_to_float::<f32>(&r).ok_or(ScalarCastError {
+            weight: self.0,
+            target: PrimitiveType::F32,
+        })
+    }
+
+    fn to_primitive(&self) -> Result<f32, ScalarCastError> {
+        let r = self.to_ratio();
+        ratio_to_float::<f32>(&r).ok_or(ScalarCastError {
+            weight: self.0.clone(),
+            target: PrimitiveType::F32,
+        })
+    }
+}
+
+impl FromPrimitive<f32> for Scalar {
+    fn from_primitive(prim: f32) -> Result<Self, PrimitiveCastError<f32>> {
+        let r = Ratio::from_float(prim).ok_or(PrimitiveCastError(prim))?;
+        Ok(Self::from_ratio(r).map_err(|_| PrimitiveCastError(prim))?)
+    }
+
+    fn from_primitive_bounded(prim: f32) -> Self {
+        let r = float_to_ratio_bounded::<f32>(prim);
+        // safe unwrap: bounded non-negative ratio r
+        Self::from_ratio(r).unwrap()
+    }
+}
+
+impl IntoPrimitive<f64> for Scalar {
+    fn into_primitive(self) -> Result<f64, ScalarCastError> {
+        let r = self.to_ratio();
+        ratio_to_float::<f64>(&r).ok_or(ScalarCastError {
+            weight: self.0,
+            target: PrimitiveType::F64,
+        })
+    }
+
+    fn to_primitive(&self) -> Result<f64, ScalarCastError> {
+        let r = self.to_ratio();
+        ratio_to_float::<f64>(&r).ok_or(ScalarCastError {
+            weight: self.0.clone(),
+            target: PrimitiveType::F64,
+        })
+    }
+}
+
+impl FromPrimitive<f64> for Scalar {
+    fn from_primitive(prim: f64) -> Result<Self, PrimitiveCastError<f64>> {
+        let r = Ratio::from_float(prim).ok_or(PrimitiveCastError(prim))?;
+        Ok(Self::from_ratio(r).map_err(|_| PrimitiveCastError(prim))?)
+    }
+
+    fn from_primitive_bounded(prim: f64) -> Self {
+        let r = float_to_ratio_bounded::<f64>(prim);
+        // safe unwrap: bounded non-negative ratio r
+        Self::from_ratio(r).unwrap()
+    }
+}
+
+/// Converts the primitive floating point value into a numerical value.
+///
+/// Maps positive/negative infinity to max/min of the primitive data type and NaN to zero.
+pub(crate) fn float_to_ratio_bounded<F: FloatCore>(f: F) -> Ratio<BigInt> {
```

Comment:
the docs don't reflect the content of the function.

also `Ratio<BigUint>` probably reflects the intention better than the current return type.

Reply:
good catch. updated docstring and refactored this a little so that the `Ratio<BigInt>` -> `Scalar` conversion also happens here, hence returning `Scalar` instead.